src/sentry/rules/actions/utils.py

```diff
@@ -0,0 +1,114 @@
+from sentry.models.environment import Environment
+from sentry.models.rule import Rule
+
+
+def get_updated_rule_data(rule: Rule) -> dict[str, Any]:
+    rule_data = dict(rule.data)
+    if rule.environment_id:
+        rule_data["environment_id"] = rule.environment_id
+    if rule.owner:
+        rule_data["owner"] = rule.owner
+    rule_data["label"] = rule.label
+    return rule_data
+
+
+def check_value_changed(
+    present_state: dict[str, Any], prior_state: dict[str, Any], key: str, word: str
+) -> str | None:
+    if present_state.get(key) != prior_state.get(key):
+        old_value = prior_state.get(key)
+        new_value = present_state.get(key)
+        return f"Changed {word} from *{old_value}* to *{new_value}*"
+    return None
+
+
+def generate_diff_labels(
+    present_state: dict[str, Any],
+    prior_state: dict[str, Any],
+    rule: Rule,
+    changed_data: DefaultDict[str, list[str]],
+    key: str,
+    statement: str,
+) -> DefaultDict[str, list[str]]:
+    for data in prior_state.get(key, []):
+        if data not in present_state.get(key, []):
+            label = generate_rule_label(rule.project, rule, data)
+            changed_data[data["id"]].append(statement.format(label))
+
+    return changed_data
+
+
+def get_changed_data(
+    rule: Rule, rule_data: dict[str, Any], rule_data_before: dict[str, Any]
+) -> dict[str, Any]:
+    """
+    Generate a list per type of issue alert rule data of what changes occurred on edit.
+    """
+    changed_data: DefaultDict[str, list[str]] = defaultdict(list)
+    changed_data = generate_diff_labels(
+        rule_data_before, rule_data, rule, changed_data, "conditions", "Added condition '{}'"
+    )
+    changed_data = generate_diff_labels(
+        rule_data, rule_data_before, rule, changed_data, "conditions", "Removed condition '{}'"
+    )
+    changed_data = generate_diff_labels(
+        rule_data_before, rule_data, rule, changed_data, "actions", "Added action '{}'"
+    )
+    changed_data = generate_diff_labels(
+        rule_data, rule_data_before, rule, changed_data, "actions", "Removed action '{}'"
+    )
+
+    frequency_text = check_value_changed(rule_data, rule_data_before, "frequency", "frequency")
+    if frequency_text:
+        changed_data["changed_frequency"].append(frequency_text)
+
+    if rule_data.get("environment_id") and not rule_data_before.get("environment_id"):
+        environment = None
+        try:
+            environment = Environment.objects.get(id=rule_data.get("environment_id"))
+        except Environment.DoesNotExist:
+            pass
+
+        if environment:
+            changed_data["environment"].append(f"Added *{environment.name}* environment")
+
+    if rule_data_before.get("environment_id") and not rule_data.get("environment_id"):
+        environment = None
+        try:
+            environment = Environment.objects.get(id=rule_data.get("environment_id"))
+        except Environment.DoesNotExist:
+            pass
```

Comment:
I'm pretty sure this is a typo -- the type checker is pointing this out as a bug when I enable model checking

this is always passing `id=None` here -- should this be using `rule_data_before["environment_id"]` instead?

Reply:
Fix here - https://github.com/getsentry/sentry/pull/73665